infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java

```diff
@@ -30,96 +39,206 @@ private StateTransfer() {
         // Hide implicit public constructor.
     }
 
+    private static final Duration MAX_RETRY_TIMEOUT = Duration.ofSeconds(10);
+
+    private static final int OVERWRITE_RETRIES = 3;
+
     /**
      * Fetch and propagate the trimMark to the new/healing nodes. Else, a FastLoader reading from
      * them will have to mark all the already trimmed entries as holes.
      * Transfer an address segment from a cluster to a set of specified nodes.
      * There are no cluster reconfigurations, hence no epoch change side effects.
      *
-     * @param layout    layout
-     * @param endpoints destination nodes
-     * @param runtime   The runtime to read the segment from
-     * @param segment   segment to transfer
+     * @param layout   layout
+     * @param endpoint destination node
+     * @param runtime  The runtime to read the segment from
+     * @param segment  segment to transfer
      */
     public static void transfer(Layout layout,
-                                Set<String> endpoints,
+                                @NonNull String endpoint,
                                 CorfuRuntime runtime,
-                                Layout.LayoutSegment segment) throws ExecutionException, InterruptedException {
+                                Layout.LayoutSegment segment) throws Exception {
 
-        if (endpoints.isEmpty()) {
-            log.debug("stateTransfer: No server needs to transfer for segment [{} - {}], " +
-                    "skipping state transfer for this segment.", segment.getStart(), segment.getEnd());
+        if (endpoint.isEmpty()) {
+            log.debug("stateTransfer: No server needs to transfer for segment [{} - {}], "
+                            + "skipping state transfer for this segment.",
+                    segment.getStart(), segment.getEnd());
             return;
         }
 
-        int batchSize = runtime.getParameters().getBulkReadSize();
+        int chunkSize = runtime.getParameters().getBulkReadSize();
+        final AtomicInteger overwriteRetries = new AtomicInteger();
```

Comment:
Why is this an AtomicInteger ?

Reply:
It's used in the lambda which requires a final variable.